### PR TITLE
Add Interval Generator

### PR DESCRIPTION
**Pull request by ciungulete**
- You can charge user monthly using Carbon addMonthsNoOverflow or you can use default generator.
- Can be used:
     - `interval => '1 month'` 
   - `interval' => [ 
'value' => 1,
'period' => 'month', 
'fixed' => false,
]`

### CODE REVIEW

**Comment by sandervanhooft** [comment]
Hi @ciungulete,

Can you

a) extract the interval generation logic into a _dedicated_ class (`PlanInterval::generate($args ...);`)?
b) the config then becomes something like:

```php
// default 
'interval' => '1 month',

// override
'interval' => [
    'class' => Laravel\Cashier\Plan\SomeSpecialIntervalGenerator::class, // or LegacyIntervalGenerator, or whatever...
    'arguments' => [
        // ...
        // all arguments here get passed in as spread
    ],
],
```

This would make it the most flexible for future modifications I think

**Comment by sandervanhooft** [comment]
Interesting to see what's evolving here. I think we now in fact try to support 3 use cases / generators.

1. string based, Cashier v1. This should be called the `LegacyCarbonBasedIntervalGenerator`. This means if people want to keep on using this after upgrading, they should update the config.
2. string based, Cashier v2 with fixed = true. This is going to be the new "default". It's easiest to use and most merchants want this interval calculation behaviour. Perhaps the best name is `V2StringBasedIntervalGenerator`? This also means that on upgrade to v2, without modification to the config this generator is automatically used.
3. Advanced `V2AdvancedIntervalGenerator`

```php
// use case 1: string based, Cashier v1. This should be called the LegacyCarbonBasedIntervalGenerator
'interval' => [
    'generator' => LegacyCarbonBasedIntervalGenerator::class,
    'value' => '1 month',
],

// use case 2
'interval' => '1 month', 

// use case 3
'interval' => [
     'generator' => DefaultIntervalGenerator::class,
     'value' => 1,
     'period' => 'month', /* day, month or year*/
     'fixed' => false, /*  If fixed is true the interval is Carbon Base Interval Generator use NoOverflow, */
 ],

```

Makes sense? 🤔

**Comment by sandervanhooft** [comment]
I'm ok with keeping the name `DefaultIntervalGenerator`, but I think the wrong class is carrying that name right now. The `V2StringBasedIntervalGenerator` could be called `DefaultIntervalGenerator`.

**Comment by ciungulete** [comment]
We have only two cases here, one with two options
```php
// 1.a (using DefaultIntervalGenerator - the old one)
'interval' => '1 month', 
//1.b (return same result like 1.a)
 'interval' => [
     'generator' => DefaultIntervalGenerator::class,
     'value' => 1,
     'period' => 'month', /* day, month or year*/
     'fixed' => false, /*  If fixed is true the interval is Carbon Base Interval Generator use NoOverflow, */
 ],

// 2.
'interval' => [
     'generator' => AdvancedIntervalGenerator::class,
     'value' => 1,
     'period' => 'month', /* day, month or year*/
     'fixed' => false, /*  If fixed is true the interval is Carbon Base Interval Generator use NoOverflow, */
 ],
```

**Comment by sandervanhooft** [comment]
The problem is that if I do a new installation I want to set

```php
'interval' => '1 month', 
```

AND have `noOverflow` enabled + base the subsequent intervals on the subscription start date. This is what most merchants want so we should make it the easiest to use / the default.

**Comment by sandervanhooft** [comment]
Another thing to be aware of:

> base the subsequent intervals on the subscription start date

this is not the whole story. A subscription can start with a trial. So we need to take the `trial_ends_at` into account too.

**Comment by ciungulete** [comment]
> A subscription can start with a trial. So we need to take the `trial_ends_at` into account too.

Yes you are right.

**Comment by ciungulete** [comment]
> The problem is that if I do a new installation I want to set
> 
> ```
> 'interval' => '1 month', 
> ```
> 
> AND have `noOverflow` enabled + base the subsequent intervals on the subscription start date. This is what most merchants want so we should make it the easiest to use / the default.

I will swap classes, but need to refactor tests... now use the old way for strings

File: src/Plan/BaseIntervalGenerator.php [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+     *
+     * @return \Carbon\Carbon
+     */
+    protected function startOfTheSubscription(Subscription  $subscription = null)
+    {
+        if (isset($subscription->trial_ends_at) && ! is_null($subscription->trial_ends_at)) {
+            return $subscription->trial_ends_at;
+        }
+
+        return $subscription_started_at = $subscription->created_at ?? now();
```
Comment (sandervanhooft) [inline src/Plan/BaseIntervalGenerator.php:22]: Can return directly

File: src/Plan/Contracts/IntervalGeneratorContract.php [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+use Laravel\Cashier\Subscription;
+
+interface IntervalGeneratorContract
+{
+    /**
+     * @param \Laravel\Cashier\Subscription $subscription
+     *
+     * @return \Carbon\Carbon
+     */
+    public function getEndOfTheNextSubscriptionCycle(Subscription  $subscription = null);
```
Comment (sandervanhooft) [inline src/Plan/Contracts/IntervalGeneratorContract.php:15]: Remove double space before `$subscription`.
Rename to `getEndOfNextSubscriptionCycle`

File: src/Plan/DefaultIntervalGenerator.php [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+     *
+     * @return \Carbon\Carbon|\Carbon\Traits\Modifiers
+     */
+    public function getEndOfTheNextSubscriptionCycle(Subscription  $subscription = null)
+    {
+        $cycle_ends_at = $subscription->cycle_ends_at ?? now();
+        $subscription_date = $this->startOfTheSubscription($subscription);
+
+        if ($this->isMonthly()) {
+            return $cycle_ends_at->addMonthsNoOverflow((int) filter_var($this->interval, FILTER_SANITIZE_NUMBER_INT))
```
Comment (sandervanhooft) [inline src/Plan/DefaultIntervalGenerator.php:35]: Nice

File: src/Plan/Plan.php [hunk 4]
```diff
@@ -28,10 +28,9 @@ class Plan implements PlanContract
     protected $amount;
 
     /**
-     * The billing frequency.
+     * The billing period.
```
Comment (sandervanhooft) [inline src/Plan/Plan.php:31]: The billing interval generator

File: config/cashier_plans.php [hunk 5]
```diff
@@ -60,7 +61,13 @@
             /**
              * The length of the billing cycle.
              */
-            'interval' => '1 month',
+            // 'interval' => 1 month or
```
Comment (sandervanhooft) [inline config/cashier_plans.php:64]: Wrap `1 month` in parantheses `"`

File: config/cashier_plans.php [hunk 6]
```diff
@@ -60,7 +61,13 @@
...
+            'interval' => [
+                'generator' => AdvancedIntervalGenerator::class,
+                'value' => 1,
+                'period' => 'month', /* day, month or year*/
+                'fixed' => false, /*  If fixed is true the interval is Carbon Base Interval Generator use NoOverflow, */
```
Comment (sandervanhooft) [inline config/cashier_plans.php:69]: Do you have a suggestion for a variable name that is more verbose, more clear about what it does?

**Review by sandervanhooft — CHANGES_REQUESTED** [review]
Almost there. Some little stuff

**Comment by sandervanhooft** [comment]
Stellar! Thanks @ciungulete !

**Pull request merged**